src/librbd/watcher/Watcher.h

```diff
@@ -0,0 +1,249 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_LIBRBD_WATCHER_H
+#define CEPH_LIBRBD_WATCHER_H
+
+#include "common/Mutex.h"
+#include "common/RWLock.h"
+#include "include/Context.h"
+#include "include/rbd/librbd.hpp"
+#include "include/rados/librados.hpp"
+#include "librbd/object_watcher/Notifier.h"
+#include "librbd/watcher/WatcherTypes.h"
+#include <set>
+#include <string>
+#include <utility>
+
+class ContextWQ;
+
+namespace librbd {
+
+template <typename T> class TaskFinisher;
+
+namespace watcher {
+
+template <typename WatcherImpl, typename Payload>
```

Comment:
Perhaps it would be a cleaner separation if ```Watcher``` used a pure virtual function to pass a payload bufferlist to the derived class. The three visitor helpers can move to a ```watcher/Utils.h``` file. That should eliminate both template params and you still get re-use.